parser.y

```diff
@@ -6034,32 +6038,118 @@ TableOptimizerHintList:
 	{
 		$$ = append($1.([]*ast.TableOptimizerHint), $2.(*ast.TableOptimizerHint))
 	}
+|	TableOptimizerHintList ',' TableOptimizerHintOpt
+	{
+		$$ = append($1.([]*ast.TableOptimizerHint), $3.(*ast.TableOptimizerHint))
+	}
 
 TableOptimizerHintOpt:
-	tidbSMJ '(' HintTableList ')'
+	hintSMJ '(' HintTableAndIndexList ')'
 	{
 		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1), Tables: $3.([]model.CIStr)}
 	}
-|	tidbINLJ '(' HintTableList ')'
+|	hintINLJ '(' HintTableAndIndexList ')'
 	{
 		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1), Tables: $3.([]model.CIStr)}
 	}
-|	tidbHJ '(' HintTableList ')'
+|	hintHJ '(' HintTableAndIndexList ')'
 	{
 		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1), Tables: $3.([]model.CIStr)}
 	}
-|	tidbHASHAGG '(' ')'
+|	hintUseIndexMerge '(' HintTableAndIndexList ')'
 	{
-		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1)}
+		$$ = &ast.TableOptimizerHint{
+			HintName: model.NewCIStr($1),
+			Tables:   $3.([]model.CIStr)[:1],
+			Indexes:  $3.([]model.CIStr)[1:],
+		}
 	}
-|	tidbSTREAMAGG '(' ')'
+|	hintUseToja '(' HintTrueOrFalse ')'
 	{
-		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1)}
+		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1), HintFlag: $3.(bool)}
+	}
+|	hintEnablePlanCache '(' HintTrueOrFalse ')'
+	{
+		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1), HintFlag: $3.(bool)}
 	}
 |	maxExecutionTime '(' NUM ')'
 	{
 		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1), MaxExecutionTime: getUint64FromNUM($3)}
 	}
+|	hintUsePlanCache '(' ')'
+	{
+		// arguments not decided yet.
+		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1)}
+	}
+|	hintQueryType '(' HintQueryType ')'
+	{
+		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1), QueryType: model.NewCIStr($3.(string))}
+	}
+|	hintMemoryQuota '(' HintMemoryQuota ')'
+	{
+		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1), MemoryQuota: $3.(uint64)}
+	}
+|	hintHASHAGG
+	{
+		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1)}
+	}
+|	hintSTREAMAGG
+	{
+		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1)}
+	}
+|	hintNoIndexMerge
+	{
+		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1)}
+	}
+|	hintReadConsistentReplica
+	{
+		$$ = &ast.TableOptimizerHint{HintName: model.NewCIStr($1)}
+	}
+
+HintTableAndIndexList:
+	Identifier
+	{
+		$$ = []model.CIStr{model.NewCIStr($1)}
+	}
+|	HintTableAndIndexList ',' Identifier
+	{
+		$$ = append($1.([]model.CIStr), model.NewCIStr($3))
+	}
+
+HintTrueOrFalse:
+	"TRUE"
+	{
+		$$ = true
+	}
+|	"FALSE"
+	{
+		$$ = false
+	}
+
+HintQueryType:
+	hintOLAP
+	{
+		$$ = $1
+	}
+|	hintOLTP
+	{
+		$$ = $1
+	}
+
+HintMemoryQuota:
+	NUM Identifier
+	{
+		// May change into MB/MiB or GB/GiB
+		switch model.NewCIStr($2).L {
+		case "m":
+			$$ = getUint64FromNUM($1)
+		case "g":
+			$$ = getUint64FromNUM($1) * 1024
+		default:
```

Comment:
Do we have better way to do this?

Reply:
Why not add a syntax for memory unit? Or you cannot raise the right warnings since the quota maybe `0M`.

Reply:
Since syntax start with number and contains a-zA-Z is difficult to handle, we may not be able to use syntax like `233M`. And in my opinion, `M` and `G` should not become a keyword, so...

Reply:
`0M` is a good case

Reply:
now we can use `MEMORY_QUOTA(233 M)`